flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperator.java

```diff
@@ -100,11 +100,14 @@ public void open() throws Exception {
 
     @Override
     public void processElement(StreamRecord<RowData> element) throws Exception {
-        if (idleTimeout > 0 && currentStatus.equals(WatermarkStatus.IDLE)) {
-            // mark the channel active
-            emitWatermarkStatus(WatermarkStatus.ACTIVE);
+        if (idleTimeout > 0) {
+            if (currentStatus.equals(WatermarkStatus.IDLE)) {
+                // mark the channel active
+                emitWatermarkStatus(WatermarkStatus.ACTIVE);
+            }
             lastRecordTime = getProcessingTimeService().getCurrentProcessingTime();
         }
```

Comment:
Thanks @pnowojski for this valuable comment!

> We can not call getProcessingTimeService().getCurrentProcessingTime() per every record, that's a too costly operation.

Good catch, I didn't notice it before.

> I think we could use a trick to count emitted records here, in the processElement. Then in WatermarkAssignerOperator#onProcessingTime you could periodically check if the processed elements count has changed and update lastRecordTime there if it did. That would loose us a little bit of accuracy, but not much. For example if processing timer is triggered ~5x more frequently than idleTimeout, the average accuracy lost would be only ~10%, which is negligible.

The solution make sense to me.

> If both idleTimeout > 0 && watermarkInterval > 0, we might need to somehow handle two timers frequencies:
> I guess we could register two different ProcessingTimeCallback with two different frequencies.
> Or we can ignore the problem and if both idleTimeout > 0 && watermarkInterval > 0, we could just have a single timer with watermarkInterval latency. This option is probably simpler, and might be good enough as usually (almost always?) watermarkInterval << idleTimeout.

In general, `watermarkInterval << idleTimeout`, but it isn't always right. Users can set them randomly, we cannot ensure how user set them. 

A workaround solution is we only register a timer with `min(idleTimeout, watermarkInterval)` latency.


> For `watermarkInterval == 0 && idleTimeout > 0`.

We can register a timer with `idleTimeout` latency.

WDYT?

Reply:
> A workaround solution is we only register a timer with min(idleTimeout, watermarkInterval) latency.

Does `idleTimeout < watermarkInterval` have any sense? Shouldn't this be disallowed even? But if you have doubts, we can support that case here.

Given that when timer is registered for value `<> watermarkInterval`, and in that case we need some special handling in the timer callback, whether it's time to emit watermark or not, maybe it makes more sense to calculate some more fancy timer interval? For example:
```
if min(watermarkInterval, idleTimeout) * 5  < max(watermarkInterval, idleTimeout)
  return min(watermarkInterval, idleTimeout)
else 
  return max(min(watermarkInterval, idleTimeout)/5, 1)
```
That would ensure our accuracy is roughly speaking always good enough?

> We can register a timer with idleTimeout latency.

Yes, that's what I had in mind :)

Reply:
> Does idleTimeout < watermarkInterval have any sense? Shouldn't this be disallowed even? But if you have doubts, we can support that case here.

Sorry, I mean we didn't forbid it, so we cannot assume it.

From your comment, I think we can limit `idleTimeout > watermarkInterval`.

Reply:
Thanks @pnowojski and @1996fanrui for the review. The approach to avoid `getProcessingTimeService().getCurrentProcessingTime()` each record seems sound to me and should be straightforward to implement.

> For watermarkInterval == 0 && idleTimeout > 0

Do we know if `watermarkInterval == 0` is a valid case where we are sure we need to periodically emit `WatermarkStatus`? My interpretation when first reading this code was that a zero value is a sentinel indicating the user isn't using watermarks. If that's the case, does the concept of "idleness" make sense anymore? Perhaps we don't emit anything for `watermarkInterval == 0`.

A [mailing list thread](https://www.mail-archive.com/user@flink.apache.org/msg40520.html) discusses some confusion in the docs on the meaning of an auto-watermark interval set to zero -- whether it means emitting a watermark each record, or not emitting them at all. 

A reply by @wuchong writes that the zero value:
> just disable **periodic** watermark emission, it doesn't mean the watermark will never be emitted.

So, there may be watermarks from elsewhere when `watermarkInterval == 0`. But does that mean we need to support idleness detection here?

The current code _would_ work in the case of a zero watermark interval, but only because of the last part of `processElement` where `advanceWatermark()` is called. That code is commented with:
> // eagerly emit watermark to avoid period timer not called (this often happens when cpu load is high)

It seems like the intent was to ensure advancement under high load, rather than support the `watermarkInterval == 0` case. Without this eager advancement, the code would never emit `Watermark`, since no timers would be scheduled & never try to detect idleness.